### ISSUE TITLE
Help "timeout" and "yield from wait(tim)"  ESP8266

### ISSUE THREAD

**Issue by dothacker495** [peterhinch/Micropython-scheduler#5]
I'm currently experimenting the accelerometer i bought recently. Tried to mix irq, poller, and pauser threads. Setting poller and pause to time out to save battery life if ever i would use it with a battery in the near future but the timeout doesn't seem to work, i also tried adding yield from wait(tim) here is a sample code

_Note: I customized my Pinblock class to cater to esp8266 irq_
```
from machine import Pin
import time
from usched import Sched, Poller, Timeout, Pinblock, wait
from LSM6DS0 import LSM6DS0
from LPS25HB import LPS25HB
def stop(fTim, objSch):
  yield fTim
  objSch.stop()

class Irq_handler(object):
  def __init__(self):
    dummy = 0
  def callback(self, irqno):
    print("Gone over THRESHOLD")

def irqtest_thread():
  han = Irq_handler()
  wf = Pinblock(Pin(10,Pin.IN),Pin.IRQ_RISING,customcallback=han.callback,timeout=None)
  count = 0
  while True:
    result = (yield wf())
    print("Interrupt recieved ",result)

def accel_thread():
  i2c = LSM6DS0()
  i2c.write_accel_settings()
  yield from wait(0.03) #Allow accelerometer to settle
  wf = Poller(i2c.read_data,(),3)#uS
  while True:
    reason = (yield wf())
    print(reason)
def pauser(objSched, thread):
    while True:
        **yield from wait(5)**
        objSched.pause(thread)
        print('Thread 2 paused: status {}'.format(status[objSched.status(thread)]))
        **yield from wait(5)**
        objSched.resume(thread)
        print('thread 2 resumed: status {}'.format(status[objSched.status(thread)]))
def test(duration = 0):
  if duration:
    print("Test interrupt for {:3d} seconds".format(duration))
  objSched = Sched()
  irq_th = objSched.add_thread(irqtest_thread())
  accel_th = objSched.add_thread(accel_thread())
  objSched.add_thread(pauser(objSched, accel_th))
  if duration:
    objSched.add_thread(stop(duration, objSched))
    
  objSched.run()

 
test()

```

The output would spam accelerometer data in mSeconds without any timout

**Comment by peterhinch** [comment]
It's not easy for me to debug this particularly as you've modified the Pinblock class in some way. If it were standard the call looks wrong to me. But perhaps the parameters are different owing to the ESP8266 firmware - which I haven't yet studied in the necessary detail. If the class were standard the call would need to look like this:
```python
wf = Pinblock(Pin(10),Pin.IN,Pin.IRQ_RISING,customcallback=han.callback,timeout=None)
```
I can't claim to understand the program. The wait on pin 10 seems unconnected to what's going on in the accelerometer thread. Does this work as expected?

The line
```python
wf = Poller(i2c.read_data,(),3)#uS
```
puzzles me. I assume i2c.read_data() returns ``None`` for most of the time, except when data is ready, otherwise the thread will constantly spam results. The us comment doesn't make obvious sense to me as the timeout is in seconds.

A general point. ``yield from wait()`` is only required for long times (> 536s) and can be replaced with ``yield 0.03`` (for example). Purely a matter of style, though. ``yield from wait()`` will work.

If you do get this working I'd very much like to see your Pinblock class adapted to the ESP8266 as I haven't got round to doing this. Good luck!

**Comment by dothacker495** [comment]
Pinblock class is working great, I also modified some parts in Waitfor class to cater esp8266 latest version of irq enable and disable

see diff log below: 
```
@@ -57,13 +57,14 @@ class Waitfor(object):
         self.interruptcount = 0                 # Set by handler, tested by triggered()
         self.roundrobin = False                 # If true reschedule ASAP
 
-    def triggered(self):                        # Polled by scheduler. Returns a priority tuple or None if not ready
+        import machine                        
         if self.irq:                            # Waiting on an interrupt
-            self.irq.disable()                  # Potential concurrency issue here (????)
+            dump = machine.disable_irq()                  # Potential concurrency issue here (????)
             numints = self.interruptcount       # Number of missed interrupts
             if numints:                         # Waiting on an interrupt and it's occurred
                 self.interruptcount = 0         # Clear down the counter
-            self.irq.enable()
+            machine.enable_irq(dump)
             if numints:
                 return (numints, 0, 0)
         if self.pollfunc:                       # Optional function for the scheduler to poll
@@ -133,21 +134,17 @@ def wait(secs):
 # Block on an interrupt from a pin subject to optional timeout. pyb specific.
 class Pinblock(Waitfor):
     initialised = False
-    def __init__(self, pin, mode, pull, customcallback = None, timeout = None):
-        if platform != 'pyboard':
-            raise ValueError('Pinblock only valid on Pyboard')
+    #pin - Pin number
+    def __init__(self, pin, mode, customcallback = None, timeout = None):
         super().__init__()
         if not Pinblock.initialised:
-            import pyb
-            import micropython
-            micropython.alloc_emergency_exception_buf(100) 
             Pinblock.initialised = True
         self.customcallback = customcallback
         if timeout is None:
             self.forever = True
         else:
             self.setdelay(timeout)
-        self.irq = pyb.ExtInt(pin, mode, pull, self.intcallback) # Porting: needs adaptation
+        self.irq = pin.irq(trigger=mode, handler=self.intcallback) # Porting: needs adaptation

```

Actually my problem lies in the accel_thread part of my code. I was just lazy to get rid of the other unrelated code not part of the problem hehe sorry. 

> "puzzles me. I assume i2c.read_data() returns None for most of the time, except when data is ready, otherwise the thread will constantly spam results." 
Ohhhh yea it spammed results. i2c.read_data() always returns a new x,y and z data 
but knowing that it spammed results, I implemented the pauser thread to pause it for at least 5-10 seconds. I used _yield 5_ before i tried _yield from wait()_ though but it still spams the results.
`
> "The us comment doesn't make obvious sense to me as the timeout is in seconds." 
Sorry about that, i forgot to remove that comment.

**Comment by peterhinch** [comment]
If i2c.read_data() always returns a new x,y and z then your accel_thread will spam results at top speed whenever the thread gets the chance to run. The only point in declaring a Poller instance is to handle the case where a user defined polling function returns data on occasion, signalling a lack of data by returning ``None``. If your function always returns data, its behaviour is identical to simply issuing ``yield`` and the thread will churn out results as fast as it can.

**Comment by dothacker495** [comment]
wouldn't pause suffice? in controlling the thread so that it won't spam data?

**Comment by peterhinch** [comment]
No, because in the times when it is running it will churn out data at full speed. Your Poller is effectively doing nothing because of the behaviour of the callback. You need to decide how to synchronise the thread which reads the accelerometer so that it performs a reading in response to some event. Alternatively you could simply ``yield interval`` to perform regular timed readings. It depends on what you're trying to achieve.

In general pausing and resuming threads is a somewhat "advanced" technique, seldom required in most applications. You need to consider each thread in isolation and determine how it is timed or synchronised.

**Comment by peterhinch** [comment]
Thinking more about this I think your accelerometer thread is hogging execution and preventing other threads from running. This is because your callback never returns ``None`` . Threads waiting on a Poller are treated as high priority resuming fast when the poller returns a value which is not ``None``. This is because the purpose of a Poller is to enable a fast response to an event. You need either to redesign your callback or abandon the Poller and synchronise the thread by other means.

**Comment by dothacker495** [comment]
Thanks for the advice. I rewrote my accel_thread, realizing that I made things very complicated when the solution was actually very simple. 

```
def accel_thread():
  yield from wait(0.03) 
  i2c = LSM6DS0()
  i2c.write_accel_settings()
  yield from wait(0.03) 
  wf = i2c.read_data,()
  while True:
    yield 15
    print(wf())
```
